app/src/cose.h

```diff
@@ -121,92 +220,463 @@ namespace scitt::cose
+  // Temporarily needed for notary_verify().
+  std::vector<uint8_t> create_sign1_tbs(
+    int cose_size,
+    std::span<const uint8_t> protected_header,
+    std::span<const uint8_t> payload)
+  {
+    // Note: This function does not return the hash of the TBS because
+    // EdDSA does not support pre-hashed messages as input.
+
+    // Sig_structure = [
+    //     context: "Signature1",
+    //     body_protected: empty_or_serialized_map,
+    //     external_aad: bstr,
+    //     payload: bstr
+    // ]
+
+    cbor::encoder ctx(cose_size + 1024);
+    QCBOREncode_OpenArray(ctx);
+
+    // context
+    QCBOREncode_AddSZString(ctx, "Signature1");
+
+    // body_protected: The protected header of the message.
+    QCBOREncode_AddBytes(ctx, cbor::from_bytes(protected_header));
+
+    // external_aad: always empty.
+    QCBOREncode_AddBytes(ctx, NULLUsefulBufC);
+
+    // payload: The payload of the message.
+    QCBOREncode_AddBytes(ctx, cbor::from_bytes(payload));
+
+    QCBOREncode_CloseArray(ctx);
+
+    return ctx.finish();
+  }
+
+  // Temporarily needed for notary_verify().
+  std::vector<uint8_t> ecdsa_sig_from_r_s(
+    const uint8_t* r,
+    size_t r_size,
+    const uint8_t* s,
+    size_t s_size,
+    bool big_endian = true)
+  {
+    OpenSSL::Unique_BIGNUM r_bn;
+    OpenSSL::Unique_BIGNUM s_bn;
+    if (big_endian)
+    {
+      OpenSSL::CHECKNULL(BN_bin2bn(r, r_size, r_bn));
+      OpenSSL::CHECKNULL(BN_bin2bn(s, s_size, s_bn));
+    }
+    else
+    {
+      OpenSSL::CHECKNULL(BN_lebin2bn(r, r_size, r_bn));
+      OpenSSL::CHECKNULL(BN_lebin2bn(s, s_size, s_bn));
+    }
+    OpenSSL::Unique_ECDSA_SIG sig;
+    OpenSSL::CHECK1(ECDSA_SIG_set0(sig, r_bn, s_bn));
+    // Ignore previous pointers, as they're now managed by ECDSA_SIG_set0
+    // https://www.openssl.org/docs/man1.1.1/man3/ECDSA_SIG_get0.html
+    (void)r_bn.release();
+    (void)s_bn.release();
+    auto der_size = i2d_ECDSA_SIG(sig, nullptr);
+    OpenSSL::CHECK0(der_size);
+    std::vector<uint8_t> der_sig(der_size);
+    auto der_sig_buf = der_sig.data();
+    OpenSSL::CHECK0(i2d_ECDSA_SIG(sig, &der_sig_buf));
+    return der_sig;
+  }
+
+  // Temporarily needed for notary_verify().
+  std::vector<uint8_t> ecdsa_sig_p1363_to_der(
+    std::span<const uint8_t> signature)
+  {
+    auto half_size = signature.size() / 2;
+    return ecdsa_sig_from_r_s(
+      signature.data(), half_size, signature.data() + half_size, half_size);
+  }
+
+  /**
+   * Extract the bstr fields from a COSE Sign1.
+   *
+   * Returns an array containing the protected headers, the payload and the
+   * signature.
+   */
+  inline std::array<std::span<const uint8_t>, 3> extract_sign1_fields(
+    std::span<const uint8_t> cose_sign1)
+  {
+    QCBORDecodeContext ctx;
+    QCBORDecode_Init(
+      &ctx, cbor::from_bytes(cose_sign1), QCBOR_DECODE_MODE_NORMAL);
+
+    QCBORDecode_EnterArray(&ctx, nullptr);
+
+    QCBORItem item;
+
+    // protected headers
+    QCBORDecode_VGetNext(&ctx, &item);
```

Comment:
All three `QCBORDecode_VGetNext` calls need to be replaced with `QCBORDecode_GetByteString`. Otherwise someone could put another type of element (eg. an integer), and create a type confusion when loading `item.val.string`.

(Using `QCBORDecode_VGetNextConsume` for the unprotected headers is fine IMO)

I realise this is old code and I even think I wrote this, but only just notice. It's much more important to fix this now this is running on envelopes that haven't been checked by t_cose.

Reply:
Done in https://github.com/microsoft/scitt-ccf-ledger/pull/90 but could you specifically review this bit please